### ISSUE TITLE
[Due for payment 2026-09-03] Update Font Styles to Match Figma Library

### ISSUE THREAD

**Issue by grgia** [Expensify/App#37503]
## Overview

The original framing — re-type the App's font values to match the Figma library — is the wrong direction and re-drifts the moment Figma or the code changes again. The design-system norm at scale is the reverse: **code is the source of truth, Figma mirrors it**, and drift is stopped by enforcement, not by periodic re-syncing.

So the goal here is not "match the numbers once." It's to establish a typography token layer that is the single source of truth, named to match Figma's semantic styles, and make raw font sizing impossible to add. Automated Figma→code sync is explicitly out of scope — it's over-engineering for a ~12-value scale that changes rarely.

## Problem

Type styles have drifted and tangled, with no mechanism to stop it:

- `fontSizeH1` and `fontSizeH2` are both `19`; `textHeadline`/`textHeadlineH1` share a size with different line heights.
- ~75 `text*` styles express ~12 real roles because color is baked into type styles (`textNormalThemeText`, `textMicroSupporting`, …), exploding combinatorially across color × size.
- 205 raw `fontSize:` and 115 raw `lineHeight:` literals bypass the scale entirely. This — not Figma changing — is where drift actually comes from.

## Scope (zero behavior change)

1. **Token file** (`src/styles/typography.ts`) — two tiers: a primitive scale (sizes, line heights, weights, families) + semantic styles named to match Figma (`body`, `bodyStrong`, `label`, `caption`, `headline`, `display`). Type only — **no color**.
2. **Alias** the existing `text*` styles to the new tokens, re-attaching color at the alias boundary so every current call-site renders byte-identically.
3. **`<Text variant>`** prop keyed to the semantic styles; deprecate the free-form `fontSize` prop.
4. **Lint rule** banning raw `fontSize`/`lineHeight` outside the token file (existing violations grandfathered via `eslint-seatbelt`).

## Out of scope

- No automated Figma→code sync pipeline.
- No big-bang migration of call-sites. The lint rule + `variant` prop pull files onto the system as they're next edited; an optional incremental sweep (one surface per PR, with screenshots) can accelerate it later.

## Example

**Foundation** — semantic styles compose the primitive scale, no color:

```ts
const textVariants = {
    body:     {...FontUtils.fontFamily.platform.EXP_NEUE,  fontSize: fontScale.body,  lineHeight: variables.fontSizeNormalHeight},
    headline: {...headlineFont, ...whiteSpace.preWrap,     fontSize: fontScale.title, lineHeight: variables.lineHeightSizeH1},
    // label, caption, bodyStrong, display…
} as const;
```

**Alias** — old names keep working, now sourced from the scale (color re-attached here):

```ts
// before
textNormalThemeText: {color: theme.text, fontSize: variables.fontSizeNormal},
textHeadlineH1: {...headlineFont, ...whiteSpace.preWrap, color: theme.heading, fontSize: variables.fontSizeXLarge, lineHeight: variables.lineHeightSizeH1},

// after — identical output
textNormalThemeText: {...textVariants.body, color: theme.text},
textHeadlineH1:      {...textVariants.headline, color: theme.heading},
```

**Call-site** — color and type decouple (`FeedbackSurvey.tsx`):

```tsx
// before
<Text style={[styles.mt1, styles.textNormalThemeText]}>{description}</Text>
// after
<Text variant="body" color={theme.text} style={styles.mt1}>{description}</Text>
```

**Lint rule** — stops the next raw literal at the door:

```
42:9  error  Raw `fontSize: 17` is not allowed. Use a <Text variant> or a token
             from src/styles/typography.ts  expensify/no-raw-typography
```

## Blast radius

- **Now:** ~6–8 files (token file + alias edits in `src/styles/index.ts` + `Text.tsx` + FontUtils map + ESLint config/rule).
- **Eventually (passive, never at once):** ~597 files reference `styles.text*` or raw font literals (~9.5% of `src`). Aliasing means none of them *need* to change; they heal one line at a time as the linter flags them.

## Needs design input

Design to confirm the semantic scale and the Figma-name → token map before the token file lands — the one part that can't be derived from code.

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c1bcd6eca2caf1d6</li>
        <li>Upwork Job ID: 1763210051994648576</li>
        <li>Last Price Increase: 2024-02-29</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @mallenexpensify</details>

**Comment by grgia** [comment]
cc @shawnborton

**Comment by dannymcclain** [comment]
You rule Georgia. 👑

**Comment by grgia** [comment]
@dannymcclain where in figma can I find these styles?

**Comment by dannymcclain** [comment]
Oh dear, just seeing your comment! If you go to the [Brand Guidelines & Product UI file](https://www.figma.com/file/qQIIaxVUdtggrh8eg8bI5B/Brand-Guidelines-%26-Product-UI?type=design&node-id=4327-4825&mode=design&t=Ipr4W9EvZpsXLNPQ-11) in Figma, you should be able to see all the text style listed in the right side toolbar/panel (make sure you don't have anything selected). And you can click on the little knobs icon to see more details about each style.

![CleanShot 2024-02-29 at 12 46 13@2x](https://github.com/Expensify/App/assets/7516624/2cd5c6a7-2548-4f35-beeb-e64dc0dab2d7)

**Comment by dannymcclain** [comment]
@grgia let me know if you need anything else for this one

**Comment by dannymcclain** [comment]
Not overdue Melvin—I don't think there's an "official" timeline for this.

**Comment by grgia** [comment]
No updates as of right now due to spring release, going to move to monthly and take a look at this when I have a spare cycle!

**Comment by dannymcclain** [comment]
Assuming this is lower priority and that Georgia is focused on upcoming releases. Correct me if I'm wrong though!

**Comment by grgia** [comment]
Yep ^ same as above

**Comment by dannymcclain** [comment]
Not prioritized right now. Lots of other projects/work going on.

**Comment by grgia** [comment]
^

**Comment by dannymcclain** [comment]
^ ^

**Comment by dannymcclain** [comment]
🤫 ^ ^ ^ Not a high priority at the moment

**Comment by dannymcclain** [comment]
@grgia have we actually done this? Or does this still need to be done? I don't even know anymore 🙈

**Comment by dannymcclain** [comment]
☝

**Comment by dannymcclain** [comment]
Not a high priority right now.

**Comment by dannymcclain** [comment]
Still not a high priority, though we really should just go ahead and do this at some point.

**Comment by grgia** [comment]
Once through F1 rush we can outsource this one

**Comment by rushatgabhane** [comment]
🏎️

**Comment by grgia** [comment]
Thank you!

**Comment by grgia** [comment]
no update

**Comment by rushatgabhane** [comment]
Please remove my assignment and assign someone else

**Comment by grgia** [comment]
@aimane-chnaif if you'd like to take a look at this, we can make a proper plan and make it external.

**Comment by aimane-chnaif** [comment]
> [@aimane-chnaif](https://github.com/aimane-chnaif) if you'd like to take a look at this, we can make a proper plan and make it external.

Sure!

**Comment by grgia** [comment]
not a priority, tho I wonder if this would be a good stress test for claude design / cowork

**Comment by grgia** [comment]
no update

**Comment by grgia** [comment]
updated description

**Comment by grgia** [comment]
<img width="405" height="307" alt="Image" src="https://github.com/user-attachments/assets/6ebdf06e-a6b3-44f1-9614-faaff5dc4e90" />



<img width="543" height="528" alt="Image" src="https://github.com/user-attachments/assets/8662ba79-4307-4996-9f7f-c5bf7266b9c8" />


@shawnborton here's the visual of the different sizes 

15 font, then 16 or 17 for emojis depending on if it's inline or menu

**Comment by shawnborton** [comment]
I think I would just make the menu emojis use 17. Thoughts?

**Comment by grgia** [comment]
Cool, yes, matched styles in the PR

**Comment by aimane-chnaif** [comment]
| Issue scope | State |
|---|---|
| 1. Token file `src/styles/typography.ts` | ✅ `src/styles/typography.ts` — `fontScale`/`lineHeightScale`/`fontFamilyScale` + 14 semantic `textVariants`, no color |
| 2. Alias `text*` styles to tokens | ⚠️ partial — see below |
| 3. `<Text variant>` + deprecate `fontSize` | ✅ `src/components/Text.tsx#L56-L75`, `fontSize` marked `@deprecated` (JSDoc only) |
| 4. Lint rule + seatbelt grandfathering | ✅ `eslint-plugin-local-rules/no-raw-typography.js`, wired at `config/eslint/eslint.config.mjs#L338`, 38 violations grandfathered across 6 files |

The design decisions from the 2026-08-11 table were all honored: H1 migration on the loading / magic-link / confirmation screens (6 files), emoji menu moved to 17 (`fontSizeEmojisWithinText`), sign-in hero / magic-code digits / eReceipt left alone, chat-count badge left as a constant.

**Remaining:**

1. **Aliasing is only partly done.** `textExtraSmall`, `textExtraSmallSupporting`, `textDoubleDecker`, `textLarge`, `textXLarge`, `textXLargeThemeText`, `textXXLarge`, `textXXXLarge`, plus the `textInputLabel/Prefix/Suffix` and `textLabelSupporting*` family still read `variables.fontSize*` directly (`src/styles/index.ts#L522-L595`). Some (e.g. `textExtraSmall` has no `lineHeight`, `textDoubleDecker` has a raw `lineHeight: 12`) can't be aliased without a visual change, so they'd need design sign-off rather than a mechanical swap.
2. **The `fontSizeH1` / `fontSizeH2` duplicate the issue opened with is still there** — both are still `19` in `src/styles/variables.ts#L68-L69`, and `fontSizeH1` survives at `src/styles/index.ts#L3455` and `src/components/ReportActionItem/MoneyReportView.tsx#L135`. `fontScale` has no token for it, so those two call-sites sit outside the system.
3. **Enforcement is narrower than "raw font sizing is impossible."** The rule only catches numeric literals in object properties and JSX attributes; `fontSize: variables.fontSizeMedium` is still legal (51 such refs outside `src/styles`), and `variables.ts` is exempt, so a new size can still be added there and used. `fontSize={variables.x}` on `<Text>` is likewise unblocked.


@grgia please check above if follow up PR is needed

**Comment by grgia** [comment]
Yes, small follow up needed. Both remaining `fontSizeH1` call-sites (`src/styles/index.ts:3471`, `MoneyReportView.tsx:135`) are 19px, same as `fontScale.h2`, so swapping them and deleting `fontSizeH1` is zero visual change. we should also widen the lint rule to catch `variables.fontSize*. The unaliased `text*` styles stay out of scope, some shift visuals so they need design sign-off first

@aimane-chnaif

**Comment by grgia** [comment]
PR is in prod with no regressions, payment is with @mallenexpensify. @aimane-chnaif any interest in the `fontSizeH1` follow up PR?

**Comment by aimane-chnaif** [comment]
> any interest in the `fontSizeH1` follow up PR?

I can work on this with Melvin

**Comment by grgia** [comment]
@aimane-chnaif sure, I opened an issue and copied over the above https://github.com/Expensify/App/issues/37503

**Comment by mallenexpensify** [comment]
### Payment Summary
Contributor+: @aimane-chnaif  due $250 via NewDot

**Comment by mallenexpensify** [comment]
I'm assuming we don't need a test case here.

**Comment by garrettmknight** [comment]
$250 approved for @aimane-chnaif
